nixos/tests/dm-verity.nix

```diff
@@ -0,0 +1,239 @@
+  imageVersion = "1-rc1";
+
+  # Use a random, but fixed root hash placeholder to allow us specifying the "real" root hash
+  # after the image is first built.
+  roothashPlaceholder = "61fe0f0c98eff2a595dd2f63a5e481a0a25387261fa9e34c37e3a4910edf32b8";
+in
+{
+  name = "verity-root";
+
+  meta.maintainers = with lib.maintainers; [ msanft ];
+
+  nodes.machine =
+    {
+      lib,
+      pkgs,
+      config,
+      modulesPath,
+      ...
+    }:
+    {
+
+      imports = [ "${modulesPath}/image/repart.nix" ];
+
+      virtualisation.directBoot.enable = false;
+      virtualisation.mountHostNixStore = false;
+      virtualisation.useEFIBoot = true;
+
+      # Disable boot loaders, as a UKI is used, which contains systemd-stub.
+      # TODO(raitobezarius): revisit this when #244907 lands
+      boot.loader.grub.enable = false;
+
+      system.image.id = imageId;
+      system.image.version = imageVersion;
+
+      fileSystems = lib.mkForce { };
+      virtualisation.fileSystems = lib.mkForce { };
+
+      # Provides 'veritysetup' to check if the verity-protected device
+      # has been mapped correctly.
+      environment.systemPackages = with pkgs; [ cryptsetup ];
+
+      boot.initrd = {
+        kernelModules = [ "overlay" ];
+        supportedFilesystems = [ "erofs" ];
+
+        systemd = {
+          enable = true;
+          dmVerity.enable = true;
+          root = "custom"; # systemd-veritysetup-generator takes care of mounting /
+
+          # We manually supply the config for systemd-tmpfiles, as the NixOS module doesn't
+          # support tmpfiles config in the initrd, but the service is running anyways.
+          contents."/etc/tmpfiles.d/mountpoints.conf".text = ''
+            d /run/etc 0755 root root -
+            d /run/etc/upper 0755 root root -
+            d /run/etc/work 0755 root root -
+            d /run/var 0755 root root -
+            d /run/var/upper 0755 root root -
+            d /run/var/work 0755 root root -
+            d /run/tmp 0755 root root -
+            d /run/tmp/upper 0755 root root -
+            d /run/tmp/work 0755 root root -
+          '';
+
+          # We directly define the mount units here, as we need to specify dependencies very
+          # granularly, and systemd-fstab-generator doesn't give us that flexibility.
+          mounts = [
+            {
+              where = "/sysroot/etc";
+              what = "overlay";
+              type = "overlay";
+              options = "lowerdir=/sysroot${config.system.build.etc}/etc,upperdir=/run/etc/upper,workdir=/run/etc/work";
+              wantedBy = [
+                "initrd-fs.target"
+                "initrd-switch-root.target"
+                "default.target"
+              ];
+              before = [ "initrd-fs.target" ];
+              after = [ "systemd-tmpfiles-setup.service" ];
+              unitConfig.RequiresMountsFor = "/sysroot/nix/store";
+              unitConfig.DefaultDependencies = false;
```

Comment:
Not that it matters for this test, but see https://github.com/NixOS/nixpkgs/pull/271326 for some discussion around `unitConfig.DefaultDependencies = false` and shutdown.